R/binning.R

```diff
@@ -98,3 +99,62 @@ bin <- function(x, y, size = 1,
                             brks[length(brks)] + mean(diff(brks))))
     brks
 }
+
+#' @title Sequence with increasing difference between elements
+#'
+#' @description
+#'
+#' `breaks_ppm` creates a sequence of numbers with increasing differences
+#' between them. Parameter `ppm` defines the amount by which the difference
+#' between values increases. The value for an element `i+1` is calculated by
+#' adding `size` to the value of element `i` and in addition also the
+#' `ppm(a, ppm)`, where `a` is the value of the element `i` plus `size`. This
+#' iterative calculation is stopped once the value of an element is larger
+#' than `to`. The last value in the result vector will thus not be equal to
+#' `to` (which is in contrast to the base [`seq()`] function) but slightly
+#' higher.
+#'
+#' A typical use case of this function would be to calculate breaks for the
+#' binning of m/z values of mass spectra. This function allows to create
+#' m/z-relative bin sizes which better represents measurement errors observed
+#' on certain mass spectrometry instruments.
+#'
+#' @param from `numeric(1)` with the value from which the sequence should start.
+#'
+#' @param to `numeric(1)` defining the upper bound of the sequence. Note that
+#'     the last value of the result will not be equal to `to` but equal to the
+#'     first number in the sequence which is larger than this value.
+#'
+#' @param by `numeric(1)` defining the constant part of the difference by which
+#'     numbers should increase.
+#'
+#' @param ppm `numeric(1)` defining the variable part of the difference by
+#'     which numbers should increase (expressed in parts-per-million of the
+#'     values).
+#'
+#' @return `numeric` with the sequence of values with increasing differences.
+#'     The returned values include `from` and `to`.
+#'
+#' @author Johannes Rainer
+#'
+#' @examples
+#'
+#' res <- breaks_ppm(20, 50, by = 1, ppm = 50)
+#' res
+#'
+#' ## difference between the values increases (by ppm)
+#' diff(res)
+breaks_ppm <- function(from = 1, to = 1, by = 1, ppm = 0) {
+    ## Maybe implement in C for higher performance.
```

Comment:
I would remove this comment. We can do it later if this function is really a bottle neck. But I assume it wouldn't be called very often.